### ISSUE TITLE
Tensor.eval() hangs with Tensorflow 1.10+

### ISSUE THREAD

**Issue by marcoancona** [tensorflow/tensorflow#23997]
**System information**
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.10.0-0-g656e7a2b34 1.10.0
- Python version: 3.6

**Describe the current behavior**
Call to `Tensor.eval()` within a custom gradient function hangs forever (see minimal example below). This seems to happen with all Tensorflow releases >=1.10.0 while it does not happen with <=1.9.0.

**Describe the expected behavior**
The tensor evaluates in the session.

**Code to reproduce the issue**
```
import numpy as np
import tensorflow as tf

SESSION = None

@tf.RegisterGradient("CustomGrad")
def _relu_grad(unused_op, grad):
    print ("Grad override fun")
    # ISSUE HERE: The following line never returns in 1.10+
    print (tf.ones((1,1)).eval(session=SESSION))
    return grad

with tf.Session() as sess:
    SESSION = sess
    g = tf.get_default_graph()
    with g.gradient_override_map({"Relu": "CustomGrad"}):
        weights = tf.random_normal((2,2)) 
        result = tf.nn.relu(weights)
        grad = tf.gradients(result, weights)[0]
        print (grad.eval())
```
Output with 1.9.0:
```
Grad override fun
[[1.]]
[[1. 1.]
 [1. 1.]]

```
Output with 1.10.0:
```
Grad override fun
--> the programs does not terminate and hangs forever here
```
